From 887242613ce67b8e8846b79e14b37b6f5ee2e955 Mon Sep 17 00:00:00 2001
From: Marcin Wojtas <mw@semihalf.com>
Date: Fri, 18 Jan 2019 16:43:21 +0100
Subject: [PATCH 0993/1051] net: mvneta: add configurable branch prediction for
 A3700 variant

This patch improves performance, using static_branch infrastructure,
depending on the SoC variant used.

Change-Id: If23945fac1fc58eecb84d4244a6ad3ba60b0b5a8
Signed-off-by: Marcin Wojtas <mw@semihalf.com>
Reviewed-on: https://sj1git1.cavium.com/3749
Tested-by: sa_ip-sw-jenkins
Reviewed-by: Stefan Chulski <Stefan.Chulski@cavium.com>
[Kevin: The original patch got from Marvell sdk10.0_19.06]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/net/ethernet/marvell/mvneta.c | 13 +++++++++----
 1 file changed, 9 insertions(+), 4 deletions(-)

diff --git a/drivers/net/ethernet/marvell/mvneta.c b/drivers/net/ethernet/marvell/mvneta.c
index 9231ff310bc3..54aebfb608e8 100644
--- a/drivers/net/ethernet/marvell/mvneta.c
+++ b/drivers/net/ethernet/marvell/mvneta.c
@@ -650,6 +650,9 @@ static int rx_header_size __read_mostly = 128;
 /* HW BM need that each port be identify by a unique ID */
 static int global_port_id;
 
+/* Branch prediction switches */
+DEFINE_STATIC_KEY_FALSE(a3700_variant);
+
 #define MVNETA_DRIVER_NAME "mvneta"
 #define MVNETA_DRIVER_VERSION "1.0"
 
@@ -2811,7 +2814,7 @@ static int mvneta_poll(struct napi_struct *napi, int budget)
 	 */
 	rx_queue = fls(((cause_rx_tx >> 8) & 0xff));
 
-	cause_rx_tx |= pp->neta_armada3700 ? pp->cause_rx_tx :
+	cause_rx_tx |= static_branch_likely(&a3700_variant) ? pp->cause_rx_tx :
 		port->cause_rx_tx;
 
 	if (rx_queue) {
@@ -2828,7 +2831,7 @@ static int mvneta_poll(struct napi_struct *napi, int budget)
 		cause_rx_tx = 0;
 		napi_complete_done(napi, rx_done);
 
-		if (pp->neta_armada3700) {
+		if (static_branch_likely(&a3700_variant)) {
 			unsigned long flags;
 
 			local_irq_save(flags);
@@ -2842,7 +2845,7 @@ static int mvneta_poll(struct napi_struct *napi, int budget)
 		}
 	}
 
-	if (pp->neta_armada3700)
+	if (static_branch_likely(&a3700_variant))
 		pp->cause_rx_tx = cause_rx_tx;
 	else
 		port->cause_rx_tx = cause_rx_tx;
@@ -4605,8 +4608,10 @@ static int mvneta_probe(struct platform_device *pdev)
 	pp->indir[0] = rxq_def;
 
 	/* Get special SoC configurations */
-	if (of_device_is_compatible(dn, "marvell,armada-3700-neta"))
+	if (of_device_is_compatible(dn, "marvell,armada-3700-neta")) {
 		pp->neta_armada3700 = true;
+		static_branch_enable(&a3700_variant);
+	}
 
 	pp->clk = devm_clk_get(&pdev->dev, "core");
 	if (IS_ERR(pp->clk))
-- 
2.17.1

